lib/logstash/inputs/dynamodb.rb

```diff
@@ -0,0 +1,341 @@
+      end
+
+      while stream_status == "ENABLING"
+        if(stream_status == "ENABLING")
+          @logger.info("Sleeping until stream is enabled")
+          sleep(1)
+        end # if stream_status == "ENABLING"
+        stream_description = dynamodb_streams_client.describeStream(AmazonDynamoDB::DescribeStreamRequest.new() \
+          .withStreamArn(stream_arn)).getStreamDescription()
+        stream_status = stream_description.getStreamStatus()
+      end # while not active
+
+      if !(stream_status == "ENABLED")
+        raise(LogStash::PluginLoadingError, "No streams are enabled")
+      end # if not active
+      @logger.info("Stream Id: " + stream_arn)
+    rescue AmazonDynamoDB::ResourceNotFoundException => rnfe
+      raise(LogStash::PluginLoadingError, rnfe.message)
+    rescue AmazonClientException => ace
+      raise(LogStash::ConfigurationError, "AWS credentials invalid or not found in the provider chain\n" + ace.message)
+    end # begin
+
+    kcl_config = KCL::KinesisClientLibConfiguration.new(@checkpointer, stream_arn, @credentials, worker_id) \
+      .withInitialPositionInStream(KCL::InitialPositionInStream::TRIM_HORIZON)
+		cloudwatch_client = nil
+    if @publish_metrics
+      cloudwatch_client = CloudWatch::AmazonCloudWatchClient.new(@credentials)
+    else
+      kclMetricsLogger = LogManager.getLogger("com.amazonaws.services.kinesis.metrics")
+      kclMetricsLogger.setAdditivity(false)
+      kclMetricsLogger.setLevel(Level::OFF)
+    end # if @publish_metrics
+    @worker = KCL::Worker.new(Logstash::Inputs::DynamoDB::LogStashRecordProcessorFactory.new(@queue), kcl_config, adapter, @dynamodb_client, cloudwatch_client)
+  end # def setup_stream
+
+  private
+  def scan(logstash_queue)
+    @logger.info("Starting scan...")
+    @logstash_writer = DynamoDBBootstrap::BlockingQueueConsumer.new(@number_of_write_threads)
+
+    @connector = DynamoDBBootstrap::DynamoDBBootstrapWorker.new(@dynamodb_client, @read_ops, @table_name, @number_of_scan_threads)
+    start_table_copy_thread
+
+    scan_queue = @logstash_writer.getQueue()
+    while true
+      event = scan_queue.take()
+      if event.getEntry().nil? and event.getSize() == -1
+        break
+      end # if event.isEmpty()
+      queue_event(@parser.parse_scan(event.getEntry(), event.getSize()), logstash_queue, @host)
+    end # while true
+  end
+
+  private
+  def stream(logstash_queue)
+    @logger.info("Starting stream...")
+    start_kcl_thread
+
+    while true
+      event = @queue.pop()
+      queue_event(@parser.parse_stream(event), logstash_queue, @host)
+    end # while true
+  end
+
+  private
+  def exit_threads
+    unless @dynamodb_scan_thread.nil?
+      @dynamodb_scan_thread.exit
+    end # unless @dynamodb_scan_thread.nil?
+
+    unless @kcl_thread.nil?
+      @kcl_thread.exit
+    end # unless @kcl_thread.nil?
+  end # def exit_threads
+
+  public
```

Comment:
As we spoke in #1, is a good practice to actually have all public methods at the beginning of the class and all private methods downstairs, in Ruby you can have something like

```
   class A
     def initialize
     end
     
     def public_method
     end
    
     private
     def private_method
     end
  end
```

everything under private is becoming private, while by default methods are private.